src/snapshot_derive/src/common.rs

```diff
@@ -0,0 +1,164 @@
+            _ => panic!("Field start/end version number must be an integer"),
+        });
+    }
+    None
+}
+
+pub(crate) fn get_end_version(attrs: &HashMap<String, syn::Lit>) -> Option<u16> {
+    if let Some(start_version) = attrs.get(END_VERSION) {
+        return Some(match start_version {
+            syn::Lit::Int(lit_int) => lit_int.base10_parse().unwrap(),
+            _ => panic!("Field start/end version number must be an integer"),
+        });
+    }
+    None
+}
+
+// Returns an attribute hash_map constructed by processing a vector of syn::Attribute.
+pub(crate) fn parse_field_attributes(
+    attributes: &Vec<syn::Attribute>,
+) -> HashMap<String, syn::Lit> {
+    let mut attrs = HashMap::new();
+
+    for nested_attr in attributes
+        .iter()
+        .flat_map(|attr| -> Result<Vec<syn::NestedMeta>, ()> {
+            if !attr.path.is_ident(ATTRIBUTE_NAME) {
+                return Ok(Vec::new());
+            }
+
+            if let Ok(syn::Meta::List(meta)) = attr.parse_meta() {
+                return Ok(meta.nested.into_iter().collect());
+            }
+
+            Ok(Vec::new())
+        })
+        .flatten()
+    {
+        if let syn::NestedMeta::Meta(nested_meta) = nested_attr {
+            if let syn::Meta::NameValue(attr_name_value) = nested_meta {
+                attrs.insert(
+                    attr_name_value.path.segments[0].ident.to_string(),
+                    attr_name_value.lit,
+                );
+            }
+        }
+    }
+
+    attrs
+}
+
+pub fn is_array(ty: &syn::Type) -> bool {
+    match ty {
+        syn::Type::Array(_) => true,
+        _ => false,
+    }
+}
+
+// Compute current struct version by finding the latest field change version.
+pub(crate) fn compute_version<T>(fields: &Vec<T>) -> u16
+where
+    T: Exists,
+{
+    let mut version = 0;
+    for field in fields {
+        version = max(version, max(field.start_version(), field.end_version()));
+    }
+    version
+}
+
+// Just checking if there are any array fields present.
+// If so, return a token stream with the vec2array macro.
+pub(crate) fn generate_deserializer_header<T>(fields: &Vec<T>) -> proc_macro2::TokenStream
+where
+    T: FieldType,
+{
+    if let Some(_) = fields.iter().find(|&field| is_array(&field.ty())) {
+        return quote! {
+            use std::convert::TryInto;
+
+            // This macro will generate a function that copies a vec to an array.
+            // We serialize arrays as vecs.
+            macro_rules! vec_to_arr_func {
+                ($name:ident, $type:ty, $size:expr) => {
+                    pub fn $name(data: &std::vec::Vec<$type>) -> [$type; $size] {
+                        let mut arr = [<$type as Default>::default(); $size];
+                        arr.copy_from_slice(&data[0..$size]);
+                        arr
+                    }
+                };
+            }
+        };
+    }
+
+    quote! {}
+}
```

Comment:
I find it hard to follow this functions since they don't belong to an object. Can we have a trait `Field` implemented by `StructField`, `EnumField`, `UnionField` ? Or a class `FieldBase` which is used by composition by all the field types enumerated before ? Or a combination of both. We could move all this logic there.

Also the Trait would probably help us to remove some duplicate logic.

Reply:
I am unsure why everything needs to belong to an object, or why this would help with readability in this case. These helper functions mostly operate on AST types. I like to keep things simple and not over engineer just be object oriented.